contributing/guide-observability.md

```diff
@@ -120,6 +120,61 @@ logs don't show up in production normally.
 For the question of what constitutes an error, errors should be actionable by a
 human. See the [Dave Cheney article] on this topic for some more discussion.
 
+## Good Errors
+
+An error message is good if it communicates
+
+1. what went wrong
+2. where it went wrong
+3. for which inputs
+
+and if it is concise without redundancies and at best displayable in one line.
+
+The general goal of an error message is to be actionable for the user who is trying
+to avoid the error in the future.
+
+Examples for bad error messages:
+- `cannot resolve package dependencies: Invalid Semantic Version`
+- `cannot parse file foo.yaml: failed to parse YAML of file foo.yaml`
+
+Examples for how to improve them:
+- `cannot resolve package dependencies in crossplane.yaml: invalid Semantic Version "main" on package "provider-kubernetes"`
+- `failed to load OCI image provider-gcp/v0.42.0: failed to unpack layer: cannot parse file: failed to parse YAML of file foo.yaml: unexpected indention in line 4`.
+
+General rules:
+1. follow the pattern `generic error cause: more details: even more details`.
+2. put values into quotes (`%q` in Golang), with the exception of
+   - resource names with namespace or resource type (e.g. `default/foo` or `providers/foo`), resource names alone are quoted
+   - kinds (e.g. `cannot find provider "provider-kubernetes`; here, the kind provider is without quotes, the provider name is with)
+   - filenames with extensions and/or path (e.g. `cannot parse foo.yaml`).
+3. add a context of e.g. filenames, object names, or values which are wrong.
+
+   Which context is added by which functions depends on the level of abstraction of the code at hand: the package loader will know the OCI image name, the YAML parser might only see a byte slice. The former adds the image name as context, some layer in-between adds the filename and the YAML parser adds the line and column in the byte stream.
+
+   Filenames should be relative to the logical working directory of the context at hand (e.g. a package root, current working directory, workspace root). Ensure that the context is clear in the error message (e.g. `failed to load package provider-kubernetes: failed to parse apis/xrd.yaml`).
+
+6. add a context just once.
+
+   Rule of thumb: if a function gets a parameter, it will also include it in the error messages, so the caller does not have to.
+   
+   E.g. a function getting a filename as input will include the filename in errors it returns.
+7. don't return multi-line errors.
+8. aggregate errors if there is clear use for the user, but not in general, i.e. fail fast is the default.
+
+   Aggregate through `kerrors.NewAggregate`. When the number of aggregated errors is unbounded,
```

Comment:
what about the recently introduced https://pkg.go.dev/errors#Join?

Reply:
Multi-line errors 😱😄

Reply:
I don't have super strong opinions on the topic tbh. I could be fine seeing some `\n` somewhere given that it's in the standard library. But I would also be fine using `kerrors.NewAggregate`, as long as we add it in `github.com/crossplane/crossplane-runtime/pkg/errors`, just to avoid having multiple error libraries.

Reply:
@jbw976 @negz opinions?

Reply:
https://goplay.tools/snippet/47cniu4DqpJ surprises me. This is really bad output, even misleading.

Reply:
vs. https://goplay.tools/snippet/6I8E-WSj4Ev 
```
something went wrong: [Lorem ipsum dolor sit amet, consectetuer adipiscing elit]
```

Reply:
An attempt: https://github.com/crossplane/crossplane-runtime/pull/478

Reply:
> ```
> something went wrong: Lorem ipsum dolor sit amet
> consectetuer adipiscing elit
> ```
I agree it doesn't look good, really unfortunate they didn't provide a way to customise the separator, similar to https://pkg.go.dev/strings#Join.

Reply:
Usually I would prefer the tooling from the stdlib, but I share the concerns about `errors.Join` joining with `\n`. I like the approach of adding a util to our own errors package.